File: _skels/python-fastapi-skel/app/wrapper_api/seed.py
"""Seed default user accounts from environment variables at startup.

Reads six env vars (three per account):

* ``USER_LOGIN``, ``USER_EMAIL``, ``USER_PASSWORD`` -- regular user
* ``SUPERUSER_LOGIN``, ``SUPERUSER_EMAIL``, ``SUPERUSER_PASSWORD`` -- admin

If *all three* vars for an account are set and no ``WrapperUser`` with
that username exists yet, the account is created. Otherwise the account
is silently skipped, making the function safe to call on every startup.
"""

from __future__ import annotations

import logging
import os

from sqlmodel import Session, select

from .db import WrapperUser
from .security import hash_password

logger = logging.getLogger(__name__)


def seed_default_accounts(session: Session) -> None:
    """Create default user and superuser accounts when env vars are set."""

    _seed_account(
        session,
        username=os.environ.get("USER_LOGIN", ""),
        email=os.environ.get("USER_EMAIL", ""),
        password=os.environ.get("USER_PASSWORD", ""),
        is_superuser=False,
        label="default user",
    )
    _seed_account(
        session,
        username=os.environ.get("SUPERUSER_LOGIN", ""),
        email=os.environ.get("SUPERUSER_EMAIL", ""),
        password=os.environ.get("SUPERUSER_PASSWORD", ""),
        is_superuser=True,
        label="superuser",
    )
    session.commit()


def _seed_account(
    session: Session,
    *,
    username: str,
    email: str,
    password: str,
    is_superuser: bool,
    label: str,
) -> None:
    if not (username and email and password):
        logger.debug("Skipping %s seed: env vars not fully set", label)
        return

    existing = session.exec(
        select(WrapperUser).where(WrapperUser.username == username)
    ).first()
    if existing is not None:
        logger.debug("Skipping %s seed: username %r already exists", label, username)
        return

    user = WrapperUser(
        username=username,
        email=email,
        hashed_password=hash_password(password),
        is_superuser=is_superuser,
    )
    session.add(user)
    logger.info("Seeded %s account: %s <%s>", label, username, email)

File: _skels/rust-actix-skel/src/seed.rs
//! Seed default user accounts from environment variables at startup.
//!
//! Reads `USER_LOGIN`, `USER_EMAIL`, `USER_PASSWORD` and
//! `SUPERUSER_LOGIN`, `SUPERUSER_EMAIL`, `SUPERUSER_PASSWORD` from the
//! process environment and inserts each account into the `users` table
//! when one with that username does not already exist.

use sqlx::sqlite::SqlitePool;

use crate::auth::hash_password;
use crate::error::ApiError;

/// Account descriptor read from env vars.
struct SeedAccount {
    username: String,
    email: String,
    password: String,
}

/// Seed default accounts (regular user + superuser) into the database.
///
/// Each account is only created when the username does not already exist,
/// making the function safe to call on every startup.
pub async fn seed_default_accounts(pool: &SqlitePool) -> Result<(), ApiError> {
    let accounts = collect_accounts();

    for acct in &accounts {
        if acct.username.is_empty() || acct.password.is_empty() {
            continue;
        }

        let existing: Option<(i64,)> =
            sqlx::query_as("SELECT id FROM users WHERE username = ?")
                .bind(&acct.username)
                .fetch_optional(pool)
                .await?;

        if existing.is_some() {
            tracing::info!(
                target: "rust_actix_skel",
                "[seed] User '{}' already exists — skipping",
                acct.username,
            );
            continue;
        }

        let hash = hash_password(&acct.password)?;
        sqlx::query(
            "INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)",
        )
        .bind(&acct.username)
        .bind(&acct.email)
        .bind(&hash)
        .execute(pool)
        .await?;

        tracing::info!(
            target: "rust_actix_skel",
            "[seed] Created default user '{}'",
            acct.username,
        );
    }

    Ok(())
}

/// Collect seed accounts from environment variables.
fn collect_accounts() -> Vec<SeedAccount> {
    vec![
        SeedAccount {
            username: std::env::var("USER_LOGIN").unwrap_or_default(),
            email: std::env::var("USER_EMAIL").unwrap_or_default(),
            password: std::env::var("USER_PASSWORD").unwrap_or_default(),
        },
        SeedAccount {
            username: std::env::var("SUPERUSER_LOGIN").unwrap_or_default(),
            email: std::env::var("SUPERUSER_EMAIL").unwrap_or_default(),
            password: std::env::var("SUPERUSER_PASSWORD").unwrap_or_default(),
        },
    ]
}
